File: bindings/node/src/zlink/contracts/messaging/received.ts
// SPDX-License-Identifier: MPL-2.0

import { SubmitError, SubmitResult } from '../errors/errors';
import type { ReplyOperation, SendOperation } from './operations';
import { RoutingId } from '../core/routing_id';
import { MessagePartsEnvelope } from './message_parts_envelope';

interface ReplyContext {
  beginReply(): ReplyOperation;
}

interface SendContext {
  beginSend(): SendOperation;
}

const replyTokenOwnerIds = new WeakMap<object, number>();
let nextReplyTokenOwnerId = 1;
let makeReplyToken!: (owner: object, value: bigint) => ReplyToken;
let tokenOwnerMatches!: (token: ReplyToken, owner: object) => boolean;
let tokenNativeValue!: (token: ReplyToken) => bigint;

/** Opaque capability required to reply to one ROUTER request. */
export class ReplyToken {
  readonly #owner: object;
  readonly #value: bigint;

  private constructor(secret: symbol, owner: object, value: bigint) {
    if (secret !== replyTokenSecret || value === 0n) {
      throw new TypeError('ReplyToken values are created by ROUTER request receive');
    }
    this.#owner = owner;
    this.#value = value;
    Object.freeze(this);
  }

  static {
    makeReplyToken = (owner: object, value: bigint): ReplyToken =>
      new ReplyToken(replyTokenSecret, owner, value);
    // Keep the capability in its private fields only. Package-private
    // readers use the same brand check instead of duplicating every token
    // in a WeakMap with a second owner/value record.
    tokenOwnerMatches = (token, owner) =>
      typeof token === 'object' && token !== null
      && #owner in token && token.#owner === owner;
    tokenNativeValue = (token) => {
      if (typeof token !== 'object' || token === null || !(#value in token)) {
        throw new TypeError('invalid ReplyToken');
      }
      return token.#value;
    };
  }

  equals(other: ReplyToken): boolean {
    return other instanceof ReplyToken
      && this.#owner === other.#owner
      && this.#value === other.#value;
  }

  hashCode(): number {
    let ownerId = replyTokenOwnerIds.get(this.#owner);
    if (ownerId === undefined) {
      ownerId = nextReplyTokenOwnerId++;
      replyTokenOwnerIds.set(this.#owner, ownerId);
    }
    const valueHash = Number((this.#value ^ (this.#value >> 32n)) & 0xffffffffn);
    return (Math.imul(ownerId, 0x9e3779b1) ^ valueHash) | 0;
  }

  toString(): 'ReplyToken' { return 'ReplyToken'; }

}

const replyTokenSecret = Symbol('ReplyToken');

/** @internal Package-private receive factory; not exported from the package root. */
export function createReplyToken(owner: object, value: bigint): ReplyToken {
  return makeReplyToken(owner, value);
}

/** @internal */
export function replyTokenOwnerMatches(token: ReplyToken, owner: object): boolean {
  return tokenOwnerMatches(token, owner);
}

/** @internal */
export function replyTokenNativeValue(token: ReplyToken): bigint {
  return tokenNativeValue(token);
}

function invalidReplyContextError(): SubmitError {
  const error = new SubmitError(SubmitResult.InvalidState, 0);
  error.message = 'reply is only valid for request-reply receive contexts';
  return error;
}

function invalidSendContextError(): SubmitError {
  const error = new SubmitError(SubmitResult.InvalidState, 0);
  error.message = 'send is only valid for received routed message contexts';
  return error;
}

/**
 * A received message envelope: its routing metadata and message parts, plus an
 * optional reply/send context.
 *
 * Owns its parts until closed. Reuse one instance across `recv` calls to avoid
 * a per-receive allocation.
 */
export class Received extends MessagePartsEnvelope {
  /** The source routing id, or null when the receive path provides none. */
  routingId: RoutingId | null;
  /** Opaque reply capability, present only for a ROUTER REQUEST receive. */
  replyToken: ReplyToken | null;
  private _replyContext: ReplyContext | null;
  private _sendContext: SendContext | null;

  /** Create an empty reusable envelope for use with `recv`. */
  constructor(...args: never[]) {
    if (args.length > 0) {
      throw new TypeError('Received reusable envelopes are created without constructor arguments');
    }
    super();
    this.routingId = null;
    this.replyToken = null;
    this._replyContext = null;
    this._sendContext = null;
  }

  /**
   * Begin a reply to this request: add parts on the returned builder, then
   * submit. Parts are consumed on a successful submit. Throws when the envelope
   * is not replyable (has no reply token).
   */
  reply(): ReplyOperation {
    if (!this.replyToken || !this._replyContext) {
      throw invalidReplyContextError();
    }
    return this._replyContext.beginReply();
  }

  /**
   * Begin a send addressed to this envelope's source route: add parts, then
   * submit. Message ownership transfers on immediate admission or after the
   * managed send takes its first back-pressure snapshot, before the Promise may
   * resolve. Throws when the envelope carries no send context.
   */
  send(): SendOperation {
    if (!this._sendContext) {
      throw invalidSendContextError();
    }
    return this._sendContext.beginSend();
  }
}

File: bindings/dotnet/src/Zlink/Contracts/Messaging/Received.cs
// SPDX-License-Identifier: MPL-2.0

namespace Systems.Zlink;

/// <summary>
///     Describes the envelope kind of a received message.
/// </summary>
public enum ReceivedMessageType
{
    /// <summary>
    ///     A plain message with no request/reply framing.
    /// </summary>
    Raw = 0,

    /// <summary>
    ///     A request that can be replied to.
    /// </summary>
    Request = 1,

    /// <summary>
    ///     A successful reply to a request.
    /// </summary>
    Reply = 2,

    /// <summary>
    ///     An error reply to a request.
    /// </summary>
    ErrorReply = 3
}

/// <summary>
///     Holds one received message envelope.
/// </summary>
/// <remarks>
///     A received envelope owns its message parts until disposed or until an API
///     explicitly transfers ownership. Reuse instances created by <c>Create</c>
///     with receive APIs that accept caller-provided storage.
/// </remarks>
public sealed partial class Received : IDisposable
{
    /// <summary>
    ///     Gets the source routing id when the receive path provides one.
    /// </summary>
    public RoutingId? RoutingId
    {
        get
        {
            return _routingIdSnapshot.GetOrCreateRoutingId(ref _routingId);
        }
    }

    /// <summary>
    ///     Gets the request sequence when this envelope can be replied to.
    /// </summary>
    public ulong? RequestSeq => _metadata?.RequestSeq;

    /// <summary>
    ///     Gets the envelope kind.
    /// </summary>
    public ReceivedMessageType MessageType { get; private set; } =
        ReceivedMessageType.Raw;

    /// <summary>
    ///     Gets the message parts owned by this envelope.
    /// </summary>
    /// <remarks>
    ///     The returned messages are disposed when this envelope is disposed unless
    ///     ownership has been explicitly transferred by another API.
    /// </remarks>
    public IReadOnlyList<Message> Parts => PartsCollection;

    /// <summary>
    ///     Gets whether the envelope currently contains exactly one message part.
    /// </summary>
    public bool IsSinglePart => _singlePart != null || PartsCollection.IsSinglePart;

    /// <summary>
    ///     Disposes the message parts owned by this envelope.
    /// </summary>
    public void Dispose()
    {
        DisposeCore();
    }

    /// <summary>
    ///     Create an empty <see cref="Received" /> for caller-provided storage.
    ///     Hand the same instance to <c>Recv(Received, ...)</c> across calls to
    ///     avoid the per-recv allocation; the binding overwrites the internal
    ///     state on each successful receive.
    /// </summary>
    public static Received Create()
    {
        return new Received();
    }

    /// <summary>
    ///     Returns the first message part without transferring ownership.
    /// </summary>
    public Message FirstPart()
    {
        return _singlePart
               ?? MessageEnvelopeParts.First(PartsCollection, nameof(Received));
    }

    /// <summary>
    ///     Returns the only message part or throws when the envelope is multipart.
    /// </summary>
    public Message SinglePartOrThrow()
    {
        return _singlePart
               ?? MessageEnvelopeParts.Single(PartsCollection, nameof(Received));
    }

    /// <summary>
    ///     Start a reply operation for request envelopes.
    /// </summary>
    /// <remarks>
    ///     The operation is valid only when <see cref="RequestSeq" /> has a value.
    /// </remarks>
    public ReplyOperation Reply()
    {
        return new ReceivedReplyOperationImpl(CaptureReplyHandler());
    }

    /// <summary>
    ///     Start a send operation addressed to the source route of this envelope.
    /// </summary>
    public SendOperation Send()
    {
        return new ReceivedSendOperationImpl(this);
    }
}
